### DISCUSSION TITLE
Lesson 13

### DISCUSSION THREAD

**Discussion by hard1k7724** [Cyfrin/foundry-full-course-cu#1823] [Q&A] [answered]

**Comment by urosognjenovic** [comment]
Here it is https://github.com/OpenZeppelin/openzeppelin-contracts/tree/master/contracts/mocks/token. It is one folder up from the folder which you provided.

**Comment by EngrPips** [comment]
Hello @Juice-07, How may we help you?

**Reply by hard1k7724** [reply]
Hi @EngrPips, Thanks but I fixed it somehow, sorry.

**Reply by EngrPips** [reply]
I love it when someone tries to unblock themselves.

**Reply by hard1k7724** [reply]
hey @EngrPips, actually I am getting an error while running : forge test --fork-url $SEPOLIA_RPC_URL 
Error - [FAIL. Reason: setup failed: environment variable "private key" not found] setUp()
How do I fix this ?

**Reply by EngrPips** [reply]
Hey @Juice-07, Can you please run the test with `-vvvv`

**Reply by hard1k7724** [reply]
Sure @EngrPips , [FAIL. Reason: setup failed: environment variable "private key" not found] setUp() (gas: 0)
Traces:
  [4253769] DSCEngineTest::setUp()
    ├─ [4129023] → new DeployDSC@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← [Return] 20511 bytes of code
    ├─ [82231] DeployDSC::run()
    │   ├─ [47344] → new MetaMultiSigWallet@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   ├─ [0] VM::envUint("private key") [staticcall]
    │   │   │   └─ ← [Revert] environment variable "private key" not found
    │   │   └─ ← [Revert] 132 bytes of code
    │   └─ ← [Revert] environment variable "private key" not found
    └─ ← [Revert] environment variable "private key" not found

Suite result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 3.91s (0.00ns CPU time)

Ran 1 test suite in 7.31s (3.91s CPU time): 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/unit/DSCEngineTest.t.sol:DSCEngineTest
[FAIL. Reason: setup failed: environment variable "private key" not found] setUp() (gas: 0)

**Reply by EngrPips** [reply]
Seem like you are trying to read private key in your deployment script, do you private key defined in your `.env` file

**Reply by hard1k7724** [reply]
Yes I have defined it, also I have run the source .env in the terminal too.

**Reply by EngrPips** [reply]
Please show your deployment script.

**Reply by hard1k7724** [reply]
//SPDX-License-Identifier:MIT

pragma solidity ^0.8.18;

import {Script} from "forge-std/Script.sol";
import {DecentralizedStableCoin} from "../src/DecentralizedStableCoin.sol";
import {DSCEngine} from "../src/DSCEngine.sol";
import {HelperConfig} from "./HelperConfig.s.sol";

contract DeployDSC is Script {
    address[] tokenAddresses;
    address[] priceFeedAddresses;

    function run()
        external
        returns (DecentralizedStableCoin, DSCEngine, HelperConfig)
    {
        HelperConfig config = new HelperConfig();

        (
            address wethUsdPriceFeed,
            address wbtcUsdPriceFeed,
            address weth,
            address wbtc,
            uint256 deployerKey
        ) = config.activeNetworkConfig();
        vm.startBroadcast(deployerKey);

        tokenAddresses = [weth, wbtc];
        priceFeedAddresses = [wethUsdPriceFeed, wbtcUsdPriceFeed];
        DecentralizedStableCoin dsc = new DecentralizedStableCoin();
        DSCEngine engine = new DSCEngine(
            tokenAddresses,
            priceFeedAddresses,
            address(dsc)
        );
        dsc.transferOwnership(address(engine));
        vm.stopBroadcast();
        return (dsc, engine, config);
    }
}
``

**Reply by EngrPips** [reply]
This shows that you read the private key in your `helperConfig,` So please show your `helperConfig`

**Reply by hard1k7724** [reply]
`//SPDX-License-Identifier:MIT

pragma solidity ^0.8.18;

import {Script} from "forge-std/Script.sol";
import {MockV3Aggregator} from "../test/mocks/MockV3Aggregator.sol";
import {ERC20Mock} from "@openzeppelin/contracts/mocks/token/ERC20Mock.sol";

contract HelperConfig is Script {
    struct NetworkConfig {
        address wethUsdPriceFeed;
        address wbtcUsdPriceFeed;
        address weth;
        address wbtc;
        uint256 deployerKey;
    }

    uint8 public constant DECIMALS = 8;
    int256 public constant ETH_USD_PRICE = 2000e8;
    int256 public constant BTC_USD_PRICE = 1000e8;
    uint256 DEFAULT_ANVIL_KEY =
        0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80;

    NetworkConfig public activeNetworkConfig;

    constructor() {
        if (block.chainid == 11155111) {
            activeNetworkConfig = getSepoliaEthConfig();
        } else {
            activeNetworkConfig = getOrCreateAnvilEthConfig();
        }
    }

    function getSepoliaEthConfig() public view returns (NetworkConfig memory) {
        return
            NetworkConfig({
                wethUsdPriceFeed: 0x694AA1769357215DE4FAC081bf1f309aDC325306, // ETH / USD
                wbtcUsdPriceFeed: 0x1b44F3514812d835EB1BDB0acB33d3fA3351Ee43,
                weth: 0xdd13E55209Fd76AfE204dBda4007C227904f0a81,
                wbtc: 0x8f3Cf7ad23Cd3CaDbD9735AFf958023239c6A063,
                deployerKey: vm.envUint("private key")
            });
    }

    function getOrCreateAnvilEthConfig() public returns (NetworkConfig memory) {
        if (activeNetworkConfig.wethUsdPriceFeed != address(0)) {
            return activeNetworkConfig;
        }
        vm.startBroadcast();
        MockV3Aggregator ethUsdPriceFeed = new MockV3Aggregator(
            DECIMALS,
            ETH_USD_PRICE
        );
        ERC20Mock wethMock = new ERC20Mock("WETH", "WETH", msg.sender, 1000e8);

        MockV3Aggregator btcUsdPriceFeed = new MockV3Aggregator(
            DECIMALS,
            BTC_USD_PRICE
        );
        ERC20Mock wbtcMock = new ERC20Mock("WBTC", "WBTC", msg.sender, 1000e8);
        vm.stopBroadcast();

        return
            NetworkConfig({
                wethUsdPriceFeed: address(ethUsdPriceFeed), // ETH / USD
                weth: address(wethMock),
                wbtcUsdPriceFeed: address(btcUsdPriceFeed),
                wbtc: address(wbtcMock),
                deployerKey: DEFAULT_ANVIL_KEY
            });
    }
}`

**Reply by EngrPips** [reply, marked as answer]
The below lies in the below code.

```js
NetworkConfig({
            wethUsdPriceFeed: 0x694AA1769357215DE4FAC081bf1f309aDC325306, // ETH / USD
            wbtcUsdPriceFeed: 0x1b44F3514812d835EB1BDB0acB33d3fA3351Ee43,
            weth: 0xdd13E55209Fd76AfE204dBda4007C227904f0a81,
            wbtc: 0x8f3Cf7ad23Cd3CaDbD9735AFf958023239c6A063,
            deployerKey: vm.envUint("private key") <@ Here to be specific
        });
```

Do you have your private key declared in your `.env` file as `private key`? I believe not. Copy whatever you name it in your `.env` file and paste it in the `vm.envUint()` and let us know if it resolves the issue.

**Reply by hard1k7724** [reply]
Hey @EngrPips thanks it is working now, I had it named as PRIVATE_KEY in my .env instead of the "private key"

**Reply by EngrPips** [reply]
Awesome, I thought you had some issues with prefixing hex with `0x`, You solved that already @Juice-07?

**Reply by hard1k7724** [reply]
Ahh yes yes I was having that issue but then I just added "0x" before the private key and it was fixed. But can you tell me if that was the correct solution to the problem @EngrPips  ?

**Reply by EngrPips** [reply]
Yeah, That was the correct solution. Keep crushing it.
